### PR TITLE
[Driver][Sparc] Default to -mcpu=v9 for 32-bit Linux/sparc64

### PR DESCRIPTION
**Pull request by rorth**
While working on supporting PR #109101 on Linux/sparc64, I was reminded that `clang -m32` still defaults to generating V8 code, although the 64-bit kernel requires a V9 CPU.

This patch corrects that on V9-only distros (Debian, Gentoo).

Tested on `sparc64-unknown-linux-gnu`, `x86_64-pc-linux-gnu`, `sparcv9-sun-solaris2.11`, and `amd64-pc-solaris2.11`.

A previous version of this patch was submitted as [[Driver][Sparc] Default to -mcpu=v9 for SparcV8 on Linux](https://reviews.llvm.org/D130688), but got stalled for 2+ years.

### CODE REVIEW

**Comment by rorth** [comment]
@glaubitz

File: clang/lib/Driver/ToolChains/Arch/Sparc.cpp
Comment (glaubitz) [reply clang/lib/Driver/ToolChains/Arch/Sparc.cpp:131]: There is the very popular SPARC Leon CPU that is used by ESA among other important customers and that uses V8 as the baseline. And Linux is supported and actively maintained on Leon.

Comment (rorth) [reply clang/lib/Driver/ToolChains/Arch/Sparc.cpp:131]: Exactly: so far I've just included distros that I know for certain are SPARC V9 only.  This can always be augmented if more come up, of course.

**Comment by MaskRay** [comment]
We try to restrict distribution differences to things like default linker options and library paths. Affecting `-mcpu=` seems very unintuitive. There are many Debian derivatives. It's weird that Debian uses -mcpu=v9 while others use -mcpu=v8. We should not increase `IsDebian` or `IsGentoo` use.

If -mcpu=v9 seems the right thing for the majority of configurations, we can bump -mcpu=v9 for all Linux and ask other, older systems to use https://clang.llvm.org/docs/UsersManual.html#configuration-files to specify `-mcpu=v8`.

**Review by thesamesam — APPROVED** [review]

**Comment by thesamesam** [comment]
(I'm fine with maskray's idea as well.)

**Comment by rorth** [comment]
> We try to restrict distribution differences to things like default linker options and library paths. Affecting `-mcpu=` seems very unintuitive. There are many Debian derivatives. It's weird that Debian uses -mcpu=v9 while others use -mcpu=v8. We should not increase `IsDebian` or `IsGentoo` use.

Ah, thanks.  That simplifies things a lot.

> If -mcpu=v9 seems the right thing for the majority of configurations, we can bump -mcpu=v9 for all Linux and ask other, older systems to use https://clang.llvm.org/docs/UsersManual.html#configuration-files to specify `-mcpu=v8`.

I'd been operating under the assumption that `clang` should work out of the box on either kind of distro.  I'll update the patch accordingly and add a release notes entry about the change.

**Comment by rorth** [comment]
Just to be certain: the revised version of the patch is ok to commit?

**Comment by glaubitz** [comment]
> Just to be certain: the revised version of the patch is ok to commit?

I think it would be good to get an ACK from at least one of the SPARC Leon maintainers.

CC @andreas-gaisler

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
> > Just to be certain: the revised version of the patch is ok to commit?
> 
> I think it would be good to get an ACK from at least one of the SPARC Leon maintainers.
> 
> CC @andreas-gaisler

Thanks for informing the maintainer. Note: this is done as a courtesy. When we bump a setting -mcpu= for good reasons, we do not add exception code like `!isVendorA() && !isVendorB()`. clangDriver is moving out of business to hard code such vendor details and requires adoption of Clang configuration files.

**Comment by rorth** [comment]
Good, that considerably simplifies non-native testing:
- The original patch ignored the issue because I had no idea how to handle it.
- The first version of this one used the hack of checking `uname -v`.  While this works, it's still wrong since this checks the host distro, not the target one.
- It occured to me that one could use the vendor component of the target triple for this purpose.  Right now, only `SUSE` is used here, but it could be extended to Debian, Gentoo and others if need be.  AFAICS, the vendor field is usually just ignored.

**Pull request merged**
